### PR TITLE
[WIP] Use fork to launch Minetest instead of system

### PR DESCRIPTION
**Pull request by rubenwardy**
- [x] Windows equivalent
- [ ] Fix warnings:

        /home/ruben/dev/NodeBoxEditor/src/minetest.cpp:170:11: warning: deprecated conversion
        from string constant to ‘char*’ [-Wwrite-strings]
           argv[1] = "--worldname";

### CODE REVIEW

File: src/minetest.cpp [hunk 1]
```diff
@@ -166,6 +163,31 @@ bool Minetest::runMod(EditorState *state, const std::string &world)
...
+	char *argv[9];
+	argv[0] = (char*)(minetest_exe).c_str();
+	argv[1] = "--worldname";
+	argv[2] = (char*)world.c_str();
+	argv[3] = "--name";
+	argv[4] = "tester";
+	argv[5] = "--address";
+	argv[6] = "";
+	argv[7] = "--go";
+	argv[8] = NULL;
```
Comment (ShadowNinja) [inline src/minetest.cpp]: You should be able to fix the warning you're getting by changing the type to `char *const argv[9]`.
Also, a nicer way to do it:
```C++
char *const argv[] = {
	minetest_exe.c_str(),
	"--worldname", world.c_str(),
	"--name", "tester",
	"--address", "",
	"--go",
	NULL,
};
```

File: src/minetest.cpp [hunk 2]
```diff
@@ -3,14 +3,18 @@
...
 #include <stdlib.h>
 #include <fstream>
 
 #ifdef _WIN32
-    #include <direct.h>
-    #define getcwd _getcwd
+	#include <direct.h>
+	#include <process.h>
+	#define getcwd _getcwd
+	#define exefilename "minetest.exe"
```
Comment (ShadowNinja) [inline src/minetest.cpp]: `EXE_FILENAME` since it's a constant?

File: src/minetest.cpp [hunk 3]
```diff
@@ -3,14 +3,18 @@
 #include "util/filesys.hpp"
 #include "FileFormat/FileFormat.hpp"
 #include "FileFormat/helpers.hpp"
+#include <stdio.h>
```
Comment (ShadowNinja) [inline src/minetest.cpp]: Is this an old debuging line?

File: src/minetest.cpp (hunk 1 above)
Comment (rubenwardy) [inline src/minetest.cpp]: Will do. It was written that way as I got an error when doing initialiser lists, so split out the lines to see which one caused it, then forgot to revert.

File: src/minetest.cpp (hunk 3 above)
Comment (rubenwardy) [inline src/minetest.cpp]: yes, will remove

File: src/minetest.cpp [hunk 4]
```diff
@@ -166,6 +163,31 @@ bool Minetest::runMod(EditorState *state, const std::string &world)
...
+
+#ifndef _WIN32
+	pid_t pid = fork();
+	if (pid == 0) {
+		chdir(minetest_dir.c_str());
+		std::cerr << "FORK: launching mt, exe = " << argv[0] << std::endl;
+		if (execvp(argv[0], argv)) {
+			std::cerr << "Couldn't run minetest!" << std::endl;
+		}
+		exit(0);
```
Comment (kaeza) [inline src/minetest.cpp]: Should use `_exit(1)` instead. Yes, the underscore is intentional. See [this answer](http://stackoverflow.com/a/2329754) for explanation.

File: src/minetest.cpp [hunk 5]
```diff
@@ -166,6 +163,31 @@ bool Minetest::runMod(EditorState *state, const std::string &world)
...
+	argv[6] = "";
+	argv[7] = "--go";
+	argv[8] = NULL;
+
+#ifndef _WIN32
+	pid_t pid = fork();
+	if (pid == 0) {
+		chdir(minetest_dir.c_str());
+		std::cerr << "FORK: launching mt, exe = " << argv[0] << std::endl;
+		if (execvp(argv[0], argv)) {
```
Comment (kaeza) [inline src/minetest.cpp]: The `if` is superfluous. If the `exec` succeeds, it doesn't return.

File: src/minetest.cpp (hunk 1 above)
Comment (rubenwardy) [inline src/minetest.cpp]: that snippet wouldn't compile, but this does but still gives warnings:

```cpp
char *const argv[] = {
    (char*)minetest_exe.c_str(),
    "--worldname", (char*)world.c_str(),
    "--name", "tester",
    "--address", "",
    "--go",
    NULL,
};
```

```
/home/ruben/dev/mt/NodeBoxEditor/src/minetest.cpp: In member function ‘bool Minetest::runMod(EditorState*, const string&)’:
/home/ruben/dev/mt/NodeBoxEditor/src/minetest.cpp:174:2: warning: deprecated conversion from string constant to ‘char*’ [-Wwrite-strings]
  };
  ^
/home/ruben/dev/mt/NodeBoxEditor/src/minetest.cpp:174:2: warning: deprecated conversion from string constant to ‘char*’ [-Wwrite-strings]
/home/ruben/dev/mt/NodeBoxEditor/src/minetest.cpp:174:2: warning: deprecated conversion from string constant to ‘char*’ [-Wwrite-strings]
/home/ruben/dev/mt/NodeBoxEditor/src/minetest.cpp:174:2: warning: deprecated conversion from string constant to ‘char*’ [-Wwrite-strings]
/home/ruben/dev/mt/NodeBoxEditor/src/minetest.cpp:174:2: warning: deprecated conversion from string constant to ‘char*’ [-Wwrite-strings]
/home/ruben/dev/mt/NodeBoxEditor/src/minetest.cpp:174:2: warning: deprecated conversion from string constant to ‘char*’ [-Wwrite-strings]
```

**Pull request merged**